keps/sig-network/3015-prefer-same-node/README.md

```diff
@@ -0,0 +1,525 @@
+>   "topologically proximate" may vary across implementations and could encompass
+>   endpoints within the same node, rack, zone, or even region.
+
+We will add a new `TrafficDistribution` value, `PreferSameZone`, which
+is defined as follows:
+
+* `PreferSameZone`: Indicates a preference for routing traffic to
+  endpoints that are in the same zone as the client. In general, the
+  proxy should always route to a same-zone endpoint if any is
+  available.
+
+And we will say that henceforth, `PreferClose` is just a deprecated
+alias for `PreferSameZone`, and never means anything other than
+"prefer same zone".
+
+Note that this does not change the existing implementation of
+`PreferClose` at all, it only removes the possibility of future
+changes.
+
+(In theory, service proxies other than kube-proxy may already have
+been implementing `PreferClose` with semantics other than the ones
+kube-proxy used. Those implementations will need to be updated when
+they add support for this KEP.)
+
+### Addition of `PreferSameNode`
+
+We will add a new `TrafficDistribution` value, `PreferSameNode`, which
+is defined as follows:
+
+* `PreferSameNode`: Indicates a preference for routing traffic to
+  endpoints that are on the same node as the client. In general, the
+  proxy should always route to a same-node endpoint if any is
+  available.
+
+We will add a new field to `discoveryv1.EndpointHints`:
+
+```golang
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+        ...
+
+	// forNodes indicates the node(s) this endpoint should be targeted by.
+	// +listType=atomic
+	ForNodes []string `json:"forNodes,omitempty" protobuf:"bytes,2,name=forNodes"`
+}
+```
+
+When updating EndpointSlices, if the EndpointSlice controller sees a
+service with `PreferSameNode` traffic distribution, then for each
+endpoint in the slice, it will add a `ForNodes` hint including the
+name of the endpoint's node. (The field is an array for future
+extensibility, but initially it will always have either 0 or 1
+elements.) In addition, it will set the `ForZones` hint as it would
+with `TrafficDistribution: PreferClose`, to allow older service
+proxies to fall back to at least same-zone behavior.
+
+Kube-proxy's `CategorizeEndpoints` function will be updated as
+follows:
+
+  - If every endpoint has a `ForNodes` hint set, and at least one
+    `ForNodes` hint includes the local node, then the set of
+    "topologically available" endpoints is the set of all endpoints
+    with a `ForNodes` hint that includes the local node.
+
+  - Otherwise, if every endpoint has a `ForZones` hint set, and at
+    least one `ForZones` hint includes the local node's zone, then the
+    set of "topologically available" endpoints is the set of all
+    endpoints with a `ForZones` hint that includes the local node's
+    zone.
+
+  - Otherwise, all endpoints are "topologically available".
```

Comment:
(nit - use italics for concept terms)
```suggestion
  - If every endpoint has a `ForNodes` hint set, and at least one
    `ForNodes` hint includes the local node, then the set of
    _topologically available_ endpoints is the set of all endpoints
    with a `ForNodes` hint that includes the local node.

  - Otherwise, if every endpoint has a `ForZones` hint set, and at
    least one `ForZones` hint includes the local node's zone, then the
    set of topologically available endpoints is the set of all
    endpoints with a `ForZones` hint that includes the local node's
    zone.

  - Otherwise, all endpoints are _topologically available_.
```